// @gmod/indexedfasta/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BgzipIndexedFasta = exports.IndexedFasta = exports.FetchableSmallFasta = exports.parseSmallFasta = void 0;
var generic_filehandle_1 = require("generic-filehandle");
var bgzipIndexedFasta_1 = __importDefault(require("./bgzipIndexedFasta"));
exports.BgzipIndexedFasta = bgzipIndexedFasta_1.default;
var indexedFasta_1 = __importDefault(require("./indexedFasta"));
exports.IndexedFasta = indexedFasta_1.default;
function parseSmallFasta(text) {
    return text
        .split('>')
        .filter(function (t) { return /\S/.test(t); })
        .map(function (entryText) {
        var _a = entryText.split('\n'), defLine = _a[0], seqLines = _a.slice(1);
        var _b = defLine.split(' '), id = _b[0], description = _b.slice(1);
        var sequence = seqLines.join('').replace(/\s/g, '');
        return {
            id: id,
            description: description.join(' '),
            sequence: sequence,
        };
    });
}
exports.parseSmallFasta = parseSmallFasta;
// memoized
var FetchableSmallFasta = /** @class */ (function () {
    function FetchableSmallFasta(_a) {
        var fasta = _a.fasta, path = _a.path;
        if (fasta) {
            this.fasta = fasta;
        }
        else if (path) {
            this.fasta = new generic_filehandle_1.LocalFile(path);
        }
        else {
            throw new Error('Need to pass fasta or path');
        }
        this.data = this.fasta.readFile().then(function (buffer) {
            var text = buffer.toString('utf8');
            return parseSmallFasta(text);
        });
    }
    FetchableSmallFasta.prototype.fetch = function (id, start, end) {
        return __awaiter(this, void 0, void 0, function () {
            var data, entry, length;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.data];
                    case 1:
                        data = _a.sent();
                        entry = data.find(function (iter) { return iter.id === id; });
                        length = end - start;
                        if (!entry) {
                            throw new Error("no sequence with id ".concat(id, " exists"));
                        }
                        return [2 /*return*/, entry.sequence.substr(start, length)];
                }
            });
        });
    };
    FetchableSmallFasta.prototype.getSequenceNames = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.data];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.map(function (entry) { return entry.id; })];
                }
            });
        });
    };
    return FetchableSmallFasta;
}());
exports.FetchableSmallFasta = FetchableSmallFasta;
